include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -103,6 +107,197 @@ void FFT(NBL_REF_ARG(Accessor) sharedmemAccessor, NBL_REF_ARG(complex_t<Scalar>)
     }
 }
 
+*/
+
+// ----------------------------------------- ABOVE MARKED FOR DELETION -----------------------------------------------------
+
+template<uint16_t K, bool Inverse, typneame Scalar, class device_capabilities=void>
+struct FFT;
+
+// For the FFT methods below, we assume:
+//      - Accessor is a global memory accessor to an array fitting 2 * _NBL_HLSL_WORKGROUP_SIZE_ elements of type complex_t<Scalar>, used to get inputs / set outputs of the FFT
+//      - SharedMemoryAccessor accesses a shared memory array that can fit _NBL_HLSL_WORKGROUP_SIZE_ elements of type complex_t<Scalar> 
+
+// 2 items per invocation forward specialization
+template<typename Scalar, class device_capabilities>
+struct FFT<2,false,device_capabilities>
+{
+    template<typename SharedMemoryAccessor>
+    static void FFT_loop(uint32_t stride, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi, uint32_t threadID, NBL_REF_ARG(MemoryAdaptor<SharedMemoryAccessor>) sharedmemAdaptor)
+    {
+        const bool topHalf = (threadID & stride) != 0;
+        vector <Scalar, 2> toTrade = topHalf ? vector <Scalar, 2>(lo.real(), lo.imag()) : vector <Scalar, 2>(hi.real(), hi.imag());
+        
+        // Block memory writes until all threads are done with previous work
+        sharedmemAdaptor.workgroupExecutionAndMemoryBarrier();
```

Comment:
you don't need the barrier on the first call to `FFT_loop` because there's no previous work (that you know about)

We have a convention in our HLSL where we assume that "accessor is always clean and doesn't alias" and if it does get reused/aliased its the responsability of the caller to ensure proper barriers are issued.

So in like with this convention you can "hoist" this barrier out of `FFT_loop` and into the `for` loop where you can check for the first iteration being performed